File: 백준/Gold/12851. 숨바꼭질 2/숨바꼭질 2.cc
#include <iostream>
# include <queue>
using namespace std;
#define MAX 100001

int N, K;
int arr[MAX] = { 0 };
int cnt, result;
queue<int> q;

void bfs(int n) {
    int cur;

    q.push(n);

    while (!q.empty()) {
        cur = q.front();
        q.pop();

        if (cur == K) {
            result = arr[cur];
            cnt += 1;
            continue;
        }

        for (int i : {cur - 1, cur + 1, cur * 2}) {
            if (i >= 0 && i < MAX && (arr[i] == 0 || arr[i] == arr[cur] + 1)) {
                arr[i] = arr[cur] + 1;
                q.push(i);
            }
        }
    }
}

int main() {
    cin >> N >> K;

    bfs(N);

    cout << result << endl << cnt;

    return 0;
}

File: 백준/Gold/12851. 숨바꼭질 2/숨바꼭질 2.py
from collections import deque

MAX = 100001

n, k = map(int, input().split())

time = [ 0 ] * (MAX)

cnt, result = 0, 0

q = deque([n])

while q:
    cur = q.popleft()

    if cur == k:
        result = time[cur]
        cnt += 1
        continue

    for i in [cur - 1, cur + 1, cur * 2]:
        if 0 <= i < MAX and (time[i] == 0 or time[i] == time[cur] + 1):
            time[i] = time[cur] + 1
            q.append(i)

print(result)
print(cnt)